### PR TITLE
Possible solution to #289 and #283

### PR DESCRIPTION
**Pull request by iViscosity**
Adds "Damagelog.markedForSlay", a table that is used when "ulx aslay(id)" is called. At the end of the round, the slays will be added. Does not modify aslayid if the target is not present.
Makes a double-check at the start of a round to make sure slain people are being set to innocent.

*Note that these are not tested, simply proposed ideas*

### CODE REVIEW

File: lua/damagelogs/shared/autoslay.lua [hunk 1]
```diff
@@ -8,8 +8,15 @@ local function CreateCommand()
...
 	local aslay = mode == 1
 
+	Damagelog.markedForSlay = {}
 	function ulx.autoslay(calling_ply, target, rounds, reason)
-		Damagelog:SetSlays(calling_ply, target:SteamID(), rounds, reason, target)
+		Damagelog.markedForSlay[ target:SteamID() ] = {
+			admin = calling_ply,
+			steamid = target:SteamID(),
+			rounds = rounds or 1,
+			reason = reason or "",
```
Comment (CSchulz) [inline lua/damagelogs/shared/autoslay.lua]: Use here `Damagelog.Autoslay_DefaultReason`

File: lua/damagelogs/shared/autoslay.lua [hunk 2]
```diff
@@ -20,7 +27,13 @@ local function CreateCommand()
 					return
 				end
 			end
-			Damagelog:SetSlays(calling_ply, target, rounds, reason, false)
+			Damagelog.markedForSlay[ target ] = {
+				admin = calling_ply,
+				steamid = target,
+				rounds = rounds or 1,
+				reason = reason or "",
```
Comment (CSchulz) [inline lua/damagelogs/shared/autoslay.lua]: Use here `Damagelog.Autoslay_DefaultReason`

File: lua/damagelogs/shared/autoslay.lua [hunk 3]
```diff
@@ -119,6 +132,16 @@ local function CreateCommand()
 end
 hook.Add("Initialize", "AutoSlay", CreateCommand)
 
+hook.Add( "TTTEndRound", "AutoSlay_CheckForMarked", function( res )
+		if not #Damagelog.markedForSlay > 0 then return end -- No need to do anything if no one is marked to be slain
+		for i = 1, #Damagelog.markedForSlay do
+			local data = Damagelog.markedForSlay[ i ]
+			Damagelog:AddSlays( data.admin, data.steamid, data.rounds, data.reason, data.target )
```
Comment (CSchulz) [inline lua/damagelogs/shared/autoslay.lua]: Any reason to switch the command from `SwitchSlays` to `AddSlays`?

File: lua/damagelogs/shared/autoslay.lua [hunk 4]
```diff
@@ -119,6 +132,16 @@ local function CreateCommand()
 end
 hook.Add("Initialize", "AutoSlay", CreateCommand)
 
+hook.Add( "TTTEndRound", "AutoSlay_CheckForMarked", function( res )
```
Comment (CSchulz) [inline lua/damagelogs/shared/autoslay.lua]: Hook should be called only on the server side.

File: lua/damagelogs/shared/autoslay.lua (hunk 3 above)
Comment (iViscosity) [reply lua/damagelogs/shared/autoslay.lua]: What do you mean `SwitchSlays`?

File: lua/damagelogs/shared/autoslay.lua (hunk 4 above)
Comment (iViscosity) [reply lua/damagelogs/shared/autoslay.lua]: Thought I wrapped it in `if SERVER then` but I guess I forgot.

File: lua/damagelogs/shared/autoslay.lua (hunk 2 above)
Comment (iViscosity) [reply lua/damagelogs/shared/autoslay.lua]: Forgot that was an option :P

File: lua/damagelogs/shared/autoslay.lua [hunk 5]
```diff
@@ -8,8 +8,15 @@ local function CreateCommand()
 	if mode != 1 and mode != 2 then return end
 	local aslay = mode == 1
 
+	Damagelog.markedForSlay = {}
 	function ulx.autoslay(calling_ply, target, rounds, reason)
-		Damagelog:SetSlays(calling_ply, target:SteamID(), rounds, reason, target)
+		Damagelog.markedForSlay[ target:SteamID() ] = {
+			admin = calling_ply,
```
Comment (Tommy228) [inline lua/damagelogs/shared/autoslay.lua]: Does that mean the slay won't work if the admin disconnects before the round end ?

Comment (iViscosity) [reply lua/damagelogs/shared/autoslay.lua]: Yeah maybe, instead of storing the player entity we could store their name instead.

Comment (CSchulz) [reply lua/damagelogs/shared/autoslay.lua]: Better store the steam id and the name for displaying.

File: lua/damagelogs/shared/autoslay.lua (hunk 3 above)
Comment (CSchulz) [reply lua/damagelogs/shared/autoslay.lua]: I meant SetSlays, don't know how it changed to SwitchSlays. Could it be possible that AddSlays is a modification in your own repository? I couldn't find anything in the original one.

File: lua/damagelogs/shared/autoslay.lua [hunk 6]
```diff
@@ -119,6 +132,18 @@ local function CreateCommand()
 end
 hook.Add("Initialize", "AutoSlay", CreateCommand)
 
+if SERVER then
+	hook.Add( "TTTEndRound", "AutoSlay_CheckForMarked", function( res )
+		if not #Damagelog.markedForSlay > 0 then return end -- No need to do anything if no one is marked to be slain
+		for i = 1, #Damagelog.markedForSlay do
+			local data = Damagelog.markedForSlay[ i ]
+			Damagelog:AddSlays( data.admin, data.steamid, data.rounds, data.reason, data.target )
```
Comment (CSchulz) [inline lua/damagelogs/shared/autoslay.lua]: AddSlays seems not to be part of the original repo.

Comment (iViscosity) [reply lua/damagelogs/shared/autoslay.lua]: [Here](https://github.com/Tommy228/TTTDamagelogs/blob/master/lua/damagelogs/server/autoslay.lua#L117)

Comment (CSchulz) [reply lua/damagelogs/shared/autoslay.lua]: Yes **Set**Slays, but you are using **Add**Slays. 😛

**Comment by iViscosity** [comment]
Oh yeah... Sorry I do a lot of these edits really late at night.

**Comment by CSchulz** [comment]
@iViscosity Can you update your PR?

**Comment by iViscosity** [comment]
Apologies for leaving this alone. I'll rewrite these and make them cleaner and make a new PR.

**Pull request closed without merging**